shared/trust/hmac.go

```diff
@@ -0,0 +1,265 @@
+package trust
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"net/http"
+	"strings"
+
+	"golang.org/x/crypto/argon2"
+
+	"github.com/canonical/lxd/shared/api"
+)
+
+// HMACVersion indicates the version used for the authorization header format.
+// This allows to define a format used by the header so that the scheme can modified
+// in future implementations without breaking already existing versions.
+// An example version can be `LXD1.0` which indicates that this is version 1.0 of the
+// LXD HMAC authentication scheme.
+// The format used after the version is dependant on the used HMACFormat:
+// Example: `Authorization: <version> <format including the HMAC>`.
+type HMACVersion string
+
+// HMACConf represents the HMAC configuration.
+type HMACConf struct {
+	HashFunc   func() hash.Hash
+	Passphrase []byte
+	Version    HMACVersion
+}
+
+// HMACFormat represents arbitrary formats to diplay and parse the actual HMAC.
+// Implementations like argon extend the format with an additional salt.
+// The Header function accepts the used version and the HMAC computed over the payload.
+// It is expected to return the final Authorization header.
+// The ParseHeader function accepts a requests header and is expected to return
+// the used version and the HMAC.
+// The Passphrase function is invoked so that the format can learn about the secret
+// used for generating the HMAC and can apply modifications such as running it through a KDF.
+// Example using argon2: `Authorization: <version> <salt>:<HMAC>`.
+type HMACFormat interface {
+	Passphrase(passphrase []byte) []byte
+	Header(version HMACVersion, hmac []byte) string
+	ParseHeader(header string) (HMACVersion, []byte, error)
+}
+
+// HMAC represents the the tooling for creating and validating HMACs.
+type HMAC struct {
+	conf   HMACConf
+	format HMACFormat
+}
+
+// HMACFormatDefault represents the default format.
+type HMACFormatDefault struct{}
```

Comment:
Please can you separate the two different hmac format options and their functions into separate files, e.g. hmac_default.go and hmac_argon2.go for clarity? You can keep the common entities in hmac.go as they are now.